### PR TITLE
Convert NVDAHelper LocalWin10 to C++/WinRT

### LINKED ISSUE #10662: NVDAHelper Windows 10, switch from WRL/CXX to C++/WinRT
We are using Windows Runtime C++ Template Library (WRL) to in operate with Windows OneCore APIs, particularly the OCR and ONeCore APIs.

From the [WRL Docs](https://docs.microsoft.com/en-gb/windows/uwp/cpp-and-winrt-apis/index):

> WRL is now superseded by C++/WinRT, a standard C++17 language projection for Windows Runtime APIs. C++/WinRT is available in the Windows 10 SDK from version 1803 onward. C++/WinRT is implemented entirely in header files, and designed to provide you with first-class access to the modern Windows API.
With C++/WinRT, you can both consume and author Windows Runtime APIs using any standards-compliant C++17 compiler. C++/WinRT typically performs better and produces smaller binaries than any other language option for the Windows Runtime. We will continue to support C++/CX and WRL, but highly recommend that new applications use C++/WinRT.

While NVDA works with WRL just fine, we should investigate a switch to WinRT as soon as new features have to be added to the Windows 10 local lib.

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Closes #10662 

### Summary of the issue:
Interaction with the Windows 10 UWP/OneCore APIs was performed using C++/CX code. This is an extension to the C++ language that which dialect is sometimes a bit different from standard C++.

C++/CX is superseded by C++/WinRT. which is a C++20 language projection. Using this results in code that is IMO much more readable to people who know C++ but are not accustomed to the CX dialect. It also uses C++ coroutines instead of the Parallel Patterns Library. Not to mention that CX is likely to be deprecated in the future.

### Description of how this pull request fixes the issue:
This pr converts all C++/CX and WRL code to its equivalent in C++/WinRT. While at it, I converted the OcSpeech and UwpOcr structs to classes that hold the speech synthesis and ocr engines as private members, ensuring that the exported functions could only access the things they need.

### Testing performed:
Tested both OCR and OneCore on a local system on Windows 10 19H1, 19H2 and 20H1. All seemed to work fine. See below for checklist

### Known issues with pull request:
None known

### Change log entry:
* Changes for developers
    + Moved the UWP/OneCore interaction layer of NVDAHelper [from C++/CX to C++/Winrt](https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/move-to-winrt-from-cx). (#10662)

### Code Review Checklist:
- [x] Pull Request description is up to date.
- [x] Unit tests.
- [x] System (end to end) tests.
- [ ] Manual testing with several versions of Windows:
  - [x] Windows 11 21H2
  - [ ] Windows 10 21H2
  - [x] Windows 10 21H1
  - [x] Windows 10 20H2
  - [x] Windows 10 2004
  - [ ] Windows 10 1909
  - [ ] Windows 10 1809 (LTSC)
  - [ ] Windows 10 1607 (LTSC)
  - [ ] Windows 10 1507 (LTSC)
  - [ ] Windows server 2022
  - [ ] Windows server 2019
  - [ ] Windows Server 2016
- [x] User Documentation.
- [x] Change log entry.
- [x] Context sensitive help for GUI changes.
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers

### CODE REVIEW

**Comment by josephsl** [comment]
CC @JCSTeh, may need to test minimum Windows 10 build supported (hopefully 17134 or later). Thanks.

**Comment by leonardder** [comment]
17134 should be supported, but I'm pretty sure that even older builds should work. I will test it against Server 2016 next week.

**Comment by Brian1Gaff** [comment]
Would any of this upset any routines used in Windows 7 or8.1?
 Brian

bglists@blueyonder.co.uk
Sent via blueyonder.
Please address personal E-mail to:-
briang1@blueyonder.co.uk, putting 'Brian Gaff'
in the display name field.
Newsgroup monitored: alt.comp.blind-users
----- Original Message -----

**Comment by dingpengyu** [comment]
Tested with nvda_snapshot_pr11768-21228 on Microsoft Windows [version 10.0.19041.572] in Simplified Chinese and no error messages appear.

**Comment by leonardder** [comment]
> Would any of this upset any routines used in Windows 7 or8.1

No. Nothing is changed for these versions of Windows.

File: nvdaHelper/archBuild_sconscript
Comment (feerrenrut) [inline nvdaHelper/archBuild_sconscript:99]: This enables co-routine support for all of the modules, I think a comment is warranted here to explain what needs this specifically.

File: nvdaHelper/localWin10/oneCoreSpeech.h
Comment (feerrenrut) [inline nvdaHelper/localWin10/oneCoreSpeech.h:28]: One thing to watch, `fire_and_forget` [swallows exceptions ](https://devblogs.microsoft.com/oldnewthing/20190320-00/?p=102345) prior to `version 1901118.3` and terminates on exception after.

File: nvdaHelper/archBuild_sconscript
Comment (feerrenrut) [reply nvdaHelper/archBuild_sconscript:99]: I think given this needs new compiler flags, and the rest of our code doesn't (yet) I'd like this PR to look at separating this code  to a separate module / dll.

**Comment by feerrenrut** [comment]
Generally I like the look of this change, but I'd like to target the 2021.1 release (and merge early for maximum testing). I'd like to mark it as a draft while we make a decision about making onecore a separate dll.

**Comment by jcsteh** [comment]
> I'd like to mark it as a draft while we make a decision about making onecore a separate dll.

Note that nvdaHelperLocalWin10 (which includes Win 10 OCR and OneCore speech) is already a separate dll. It previously had a separate SCons environment which allowed compiler flags to be tweaked, but this PR removes that. If you're worried about coroutine support being enabled everywhere, I'd suggest reinstating that separate SCons environment and moving the /await flag to that.

Note that given the highly asynchronous nature of WinRT in general, I think most things in nvdaHelperLocalWin10 are probably going to need await sooner or later, so I think it's reasonable to enable this for the whole of nvdaHelperLocalWin10 (but not necessarily other NVDA dlls).

All of that said, I can think of other places in nvdaHelper where coroutines could be useful, but it's certainly a riskier change that way.

File: nvdaHelper/archBuild_sconscript
Comment (leonardder) [reply nvdaHelper/archBuild_sconscript:99]: The code is already in a separate dll.
Note that before this pr, we had a separate win 10 environment that added compiler flags to support C++/CX, but I think changing parameters at the sconscript level is ok.
Having said that, I made /await global as I think there's potential in using co-routines more widely in our codebase. For example, @michaelDCurran used APCs in #11431 and #11425, but co-routines could have been appropriate here as well.

File: nvdaHelper/localWin10/oneCoreSpeech.h
Comment (leonardder) [reply nvdaHelper/localWin10/oneCoreSpeech.h:28]: I believe we catch everything we caught before, so this shouldn't hurt us I believe.

File: nvdaHelper/archBuild_sconscript
Comment (feerrenrut) [inline nvdaHelper/archBuild_sconscript:150]: Please add "for winrt usage" or something more specific. I haven't yet done a thorough review of this so I'm not sure I can make a precise suggestion yet. Generally, if I encountered this while trying to split up the code in this DLL or merge with another, or perhaps copy the approach for a new DLL, then I'd want to know which code / features depend on this flag and why it is necessary.

File: nvdaHelper/archBuild_sconscript
Comment (feerrenrut) [reply nvdaHelper/archBuild_sconscript]: Yes, sorry I missed that.

`/await` may be useful in other areas too. But I'd like that to decided on a case by case basis. Given this hasn't actually been finalized in the C++ standard yet (but included in draft for C++20), we should be cautious about how we use it. I think using it via WinRT should be fine, they will have handle any adjustments as the compiler adjusts. This recent blog post talks about the differences between `/await` and C++20: https://devblogs.microsoft.com/cppblog/c-coroutines-in-visual-studio-2019-version-16-8/

 It would be useful to call this out (and the blog) on the PR description.

File: nvdaHelper/localWin10/oneCoreSpeech.h
Comment (feerrenrut) [reply nvdaHelper/localWin10/oneCoreSpeech.h:28]: You'll have to confirm which behaviour we get. If we are using post `version 1901118.3` then an exception will kill the app (via a call to terminate). Or perhaps you are saying that inside `speak` all exceptions are caught?

File: nvdaHelper/archBuild_sconscript
Comment (feerrenrut) [reply nvdaHelper/archBuild_sconscript]: After reading that blog post more carefully, I think it is worth considering using `/std:c++latest` rather than `/await`.

**Comment by lukaszgo1** [comment]
Hi @leonardder 
I've tried a build from this PR on my Windows 10 1511 box, and unfortunately OneCore driver fails to load with the following, not very useful IMHO, log:

```
IO - inputCore.InputManager.executeGesture (01:00:22.585) - winInputHook (956):
Input: kb(desktop):enter
DEBUG - autoSettingsUtils.autoSettings.AutoSettings._registerConfigSaveAction (01:00:22.602) - MainThread (4764):
registering pre_configSave action: <class 'synthDrivers.oneCore.SynthDriver'>
DEBUGWARNING - NVDAHelperLocal (01:00:22.604) - MainThread (4764):
Thread 4764, build\x86\localWin10\oneCoreSpeech.cpp, OcSpeech::OcSpeech, 43:
AppendedSilence not supported

ERROR - synthDriverHandler.setSynth (01:00:22.604) - MainThread (4764):
setSynth
Traceback (most recent call last):
  File "synthDriverHandler.pyc", line 448, in setSynth
  File "synthDriverHandler.pyc", line 420, in getSynthInstance
  File "synthDriverHandler.pyc", line 303, in initSettings
  File "synthDrivers\oneCore.pyc", line 266, in _get_rate
OSError: exception: access violation reading 0x00000000
```


While W10 1511 is quite old, unsupported and therefore not very important the fact that this code fails there suggests to me that it would also fail on 1507 LTSB which is still supported until October 2025 so I believe we need to support  it too.

In general I'd like to push back a little on this. Having more readable code is of course important but since it brings no user visible benefits merging this PR seems quite risky to me. We had similar discussion in the past on #10405 which quite similarly to this one just modernized syntax without user visible changes and all parties involved there decided that it should be closed.
IMHO it makes sense to delay reviewing/merging this until one of the following happens:
- Microsoft announces that C++/CX is indeed going to be deprecated - for now when they intent to do this is anyone's guess.
-  Some sort of a bug  is found in the localWin10 code which would require a significant rewrite to fix anyway.

BTW Were you able to test this on Server 2016?

**Comment by leonardder** [comment]
> In general I'd like to push back a little on this. Having more readable code is of course important but since it brings no user visible benefits merging this PR seems quite risky to me. 

If properly tested, I'm not sure why this code introduces risk. We just will have to make sure that we test all of it in the full range of still supported Windows 10 builds.

> We had similar discussion in the past on #10405 which quite similarly to this one just modernized syntax without user visible changes and all parties involved there decided that it should be closed.

There's a huge difference between Python and C++ in that C++ compiles all code before running it, which tackles most of the risk IMO. Changing Python code has a way bigger risk of experiencing regressions at run time.

> IMHO it makes sense to delay reviewing/merging this until one of the following happens:
> 
>     * Microsoft announces that C++/CX is indeed going to be deprecated - for now when they intent to do this is anyone's guess.

True. However, they already highly recomment C++/WinRT at this point.

I'd like to add that coroutine suppot in C++/WinRT is currently based on the experimental support in C++17. The current WinRT bindings in the Windows 10 SDK don't support C++20 style coroutines. We could consider this another reason to wait for this to be merged.

**Comment by lukaszgo1** [comment]
Unfortunately latest build from this { PR still crashes on Windows 10 1511. Log is as follows:

DEBUG - autoSettingsUtils.autoSettings.AutoSettings._registerConfigSaveAction (20:56:04.880) - MainThread (1812):
registering pre_configSave action: <class 'synthDrivers.oneCore.SynthDriver'>
DEBUGWARNING - NVDAHelperLocal (20:56:04.892) - MainThread (1812):
Thread 1812, build\x86\localWin10\oneCoreSpeech.cpp, OcSpeech::OcSpeech, 46:
AppendedSilence not supported

ERROR - synthDriverHandler.setSynth (20:56:04.897) - MainThread (1812):
setSynth
Traceback (most recent call last):
  File "synthDriverHandler.pyc", line 448, in setSynth
  File "synthDriverHandler.pyc", line 420, in getSynthInstance
  File "synthDriverHandler.pyc", line 303, in initSettings
  File "synthDrivers\oneCore.pyc", line 268, in _get_rate
OSError: exception: access violation reading 0x00000000

**Comment by lukaszgo1** [comment]
> If properly tested, I'm not sure why this code introduces risk. We just will have to make sure that we test all of it in the full range of still supported Windows 10 builds.

Most of testing coverage we can get comes from Alpha users and Server versions of Windows / LTSB editions of Windows 10 are used mostly in corporate environments so I don't think many people are willing/able to use test builds in such circumstances.


> There's a huge difference between Python and C++ in that C++ compiles all code before running it, which tackles most of the risk IMO. Changing Python code has a way bigger risk of experiencing regressions at run time.

Its obviously true when considering risks which comes from syntactic errors, but when it comes to regression testing our Python code has a lot of unit tests which simply does not exists for the C++ part.

In essence with changes like this it is more than likely that some kind of regression will slip through especiallly for older Win 10 versions so we're trading not so pretty but certainly working (it has been written back when old versions of Win 10 were a current one so it supports them well) code for a new one which is written in a modern way but may or ma not  work.

**Comment by leonardder** [comment]
Regardless what we do, I think I'd like to go to the bottom of this. I might be able to test on server 2016 later on.

Could you provide output of the following on 1511? `NVDAHelper.getHelperLocalWin10Dll().ocSpeech_supportsProsodyOptions()`

**Comment by lukaszgo1** [comment]
> Could you provide output of the following on 1511? `NVDAHelper.getHelperLocalWin10Dll().ocSpeech_supportsProsodyOptions()`

On the latest Alpha it returns 0, whereas with the build from this PR 1631488 is returned.

**Comment by leonardder** [comment]
Wow, there's some real weirdness going on there indeed.

**Comment by leonardder** [comment]
It looks like IsApiContractPresent panics if you provide it a regular string that's implicitly converted to a hstring. Providing it a hstring seems to fix it. @lukaszgo1 could you verify?

**Comment by lukaszgo1** [comment]
Yes, OneCore works correctly using latest build on 1511.

**Comment by MarcoZehe** [comment]
@leonardder Is this still something you want to follow through with?

**Comment by leonardder** [comment]
Sure. I really believe that, though this pr does not change anything in behavior, it really improves readability and durability of this code. Microsoft has plans to eventually merge C++/WinRT with their new C++/WIN32 project which NVDA in my opinion should heavily rely on in the future.
I think there'might be one reason to wait for this to be merged as noted in https://github.com/nvaccess/nvda/pull/11768#issuecomment-740809514:

> I'd like to add that coroutine suppot in C++/WinRT is currently based on the experimental support in C++17. The current WinRT bindings in the Windows 10 SDK don't support C++20 style coroutines.

I don't think it's a difference in behaviour though.

**Comment by leonardder** [comment]
Probably best to delay this until Windows 11 is released and we have a new SDK with up to date WinRT

**Comment by lijianwei2019** [comment]
Windows 11 has been released

**Comment by seanbudd** [comment]
I've triggered a new build now that AppVeyor has updated the Win SDK

**Comment by leonardder** [comment]
> Probably best to delay this until Windows 11 is released and we have a new SDK with up to date WinRT

This is now the case.
Note that I only enabled C++20 for NVDAHelperWin10, not for all the code of NVDAHelper, as that imposes some incompatibilities.

**Comment by lukaszgo1** [comment]
> If properly tested, I'm not sure why this code introduces risk. We just will have to make sure that we test all of it in the full range of still supported Windows 10 builds.

Would you be able to update the initial description to talk about what testing you did in that regard?

File: nvdaHelper/localWin10/uwpOcr.cpp
Comment (feerrenrut) [inline nvdaHelper/localWin10/uwpOcr.cpp:76]: Please use `nullptr` not NULL

File: nvdaHelper/localWin10/oneCoreSpeech.cpp
Comment (feerrenrut) [inline nvdaHelper/localWin10/oneCoreSpeech.cpp:64]: Please add a comment here about exceptions escaping this method will result in a call to `terminate`: https://devblogs.microsoft.com/oldnewthing/20190320-00/?p=102345

**Comment by leonardder** [comment]
> Would you be able to update the initial description to talk about what testing you did in that regard?

I just did. I can test on Server 2019 and Server 2016 myself. Having said that, if it works on the newest and oldest versions of Windows 10, I guess it's safe to assume that it works on the others too.

File: nvdaHelper/localWin10/oneCoreSpeech.h
Comment (feerrenrut) [reply nvdaHelper/localWin10/oneCoreSpeech.h:28]: Perhaps move the contents of `speak` into a private helper, then in the public `fire_and_forget` wrapper catch all exceptions and log them.

**Review by feerrenrut — CHANGES_REQUESTED** [review]
We'll have to wait until we start the 2022.1 dev cycle before merging this.

**Comment by feerrenrut** [comment]
Blocked until the 2022.1 dev cycle starts.

File: nvdaHelper/localWin10/oneCoreSpeech.h
Comment (leonardder) [reply nvdaHelper/localWin10/oneCoreSpeech.h:28]: I'm not sure whether I follow you correctly. Moving the contents of speak to a private helper still requires it to be an coroutine. Wouldn't it be enough to wrap just everything in a try block?

**Review by feerrenrut — CHANGES_REQUESTED** [review]
Ideally, I'd like as much of the algorithm as possible decoupled with it's execution policy. Please separate the parts of the code that could be independent pieces and then combine them with a method that knows about the threading approach.

This makes it easier to reason about the execution policy (and the algorithm) without having to consider both at once.

File: source/comInterfaces_sconscript
Comment (lukaszgo1) [inline source/comInterfaces_sconscript:93]: What is this change about?

Comment (leonardder) [reply source/comInterfaces_sconscript:93]: Ugh, that must have been something leaking in from my work on #12953. I'll revert that.

File: nvdaHelper/localWin10/uwpOcr.cpp
Comment (feerrenrut) [inline nvdaHelper/localWin10/uwpOcr.cpp:86]:
```suggestion
	Buffer buf { numBytes };
```

File: nvdaHelper/localWin10/uwpOcr.cpp
Comment (feerrenrut) [inline nvdaHelper/localWin10/uwpOcr.cpp:43]: This isn't very safe (potential memory leaks, loss of ownership), this should use a smart pointer instead.
We have the opportunity to change this design for the 2022.1 release, or we can maintain the design and (I think) return a ccomPtr to Python.

File: nvdaHelper/localWin10/oneCoreSpeech.cpp
Comment (feerrenrut) [inline nvdaHelper/localWin10/oneCoreSpeech.cpp:81]:
```suggestion
		Buffer buffer { size };

```

File: nvdaHelper/localWin10/oneCoreSpeech.cpp
Comment (feerrenrut) [inline nvdaHelper/localWin10/oneCoreSpeech.cpp:69]: I don't see why this should be a shared pointer?

File: nvdaHelper/localWin10/oneCoreSpeech.cpp
Comment (feerrenrut) [inline nvdaHelper/localWin10/oneCoreSpeech.cpp:48]: As per OCR, let's reconsider this approach.

File: nvdaHelper/localWin10/oneCoreSpeech.cpp
Comment (feerrenrut) [reply nvdaHelper/localWin10/oneCoreSpeech.cpp:64]: Can you still add a comment here about why we catch all exceptions.

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: nvdaHelper/localWin10/oneCoreSpeech.cpp
Comment (feerrenrut) [inline nvdaHelper/localWin10/oneCoreSpeech.cpp:65]: Typo 'al' -> 'all'
```suggestion
	// Ensure we catch all exceptions in this method,

```

File: nvdaHelper/localWin10/uwpOcr.cpp
Comment (feerrenrut) [inline nvdaHelper/localWin10/uwpOcr.cpp:52]: Typo 'al' -> 'all'
```suggestion
	// Ensure we catch all exceptions in this method,

```

**Review by feerrenrut — CHANGES_REQUESTED** [review]
Let's leave the other more complicated design changes for a follow up. With these minor fixes I propose we merge.

File: nvdaHelper/localWin10/oneCoreSpeech.cpp
Comment (feerrenrut) [reply nvdaHelper/localWin10/oneCoreSpeech.cpp:48]: We can follow up on this via new issue: https://github.com/nvaccess/nvda/issues/13253

File: nvdaHelper/localWin10/uwpOcr.cpp
Comment (feerrenrut) [reply nvdaHelper/localWin10/uwpOcr.cpp:43]: We can follow up on this via new issue: https://github.com/nvaccess/nvda/issues/13253

File: nvdaHelper/localWin10/uwpOcr.cpp
Comment (feerrenrut) [reply nvdaHelper/localWin10/uwpOcr.cpp]: We can follow up on this via new issue: https://github.com/nvaccess/nvda/issues/13253

**Review by feerrenrut — APPROVED** [review]

**Review by feerrenrut — APPROVED** [review]

**Pull request merged**
